dot/rpc/websocket.go

```diff
@@ -0,0 +1,161 @@
+package rpc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ChainSafe/gossamer/dot/rpc/modules"
+	log "github.com/ChainSafe/log15"
+	"github.com/gorilla/mux"
+	"github.com/gorilla/rpc/v2"
+	"github.com/gorilla/websocket"
+)
+
+// HTTPServerWS gateway for WebSocket RPC server
+type HTTPServerWS struct {
+	rpcServer    *rpc.Server // Actual RPC call handler
+	serverConfig *HTTPServerConfigWS
+}
+
+// HTTPServerConfigWS configures the WebSocket HTTPServer
+type HTTPServerConfigWS struct {
+	HTTPServerConfig *HTTPServerConfig
+	RPCPort          uint32
+}
+
+// NewHTTPServerWS creates a new http server and registers an associated rpc server
+func NewHTTPServerWS(cfg *HTTPServerConfigWS) *HTTPServerWS {
+	server := &HTTPServerWS{
+		rpcServer:    rpc.NewServer(),
+		serverConfig: cfg,
+	}
+
+	server.RegisterModules(cfg.HTTPServerConfig.Modules)
+	return server
+}
+
+// RegisterModules registers the RPC services associated with the given API modules
+func (h *HTTPServerWS) RegisterModules(mods []string) {
+
+	for _, mod := range mods {
+		log.Debug("[rpc] Enabling rpc module", "module", mod)
+		var srvc interface{}
+		switch mod {
+		case "system":
+			srvc = modules.NewSystemModule(h.serverConfig.HTTPServerConfig.NetworkAPI)
+		case "author":
+			srvc = modules.NewAuthorModule(h.serverConfig.HTTPServerConfig.CoreAPI, h.serverConfig.HTTPServerConfig.TransactionQueueAPI)
+		case "chain":
+			srvc = modules.NewChainModule(h.serverConfig.HTTPServerConfig.BlockAPI)
+		default:
+			log.Warn("[rpc] Unrecognized module", "module", mod)
+			continue
+		}
+
+		err := h.rpcServer.RegisterService(srvc, mod)
+
+		if err != nil {
+			log.Warn("[rpc] Failed to register module", "mod", mod, "error", err)
+		}
+	}
+}
+
+// Start registers the rpc handler function and starts the server listening on `h.port`
+func (h *HTTPServerWS) Start() error {
+	// use our DotUpCodec which will capture methods passed in json as _x that is
+	//  underscore followed by lower case letter, instead of default RPC calls which
+	//  use . followed by Upper case letter
+	h.rpcServer.RegisterCodec(NewDotUpCodec(), "application/json")
+	r := mux.NewRouter()
+	r.Handle("/", h.rpcServer)
+
+	r.Use(h.websocketHandler)
```

Comment:
I'm guessing you can't get the router to use multiple handlers which is why you need two servers? in that case, this seems like a good approach

Reply:
Right, this is my concern, but at this point I think it's the best way to go.